merlin/models/tf/models/base.py

```diff
@@ -812,60 +812,172 @@ def call_train_test(
+        target = tf.cast(target, prediction.dtype)
+
+        # If targets are scalars (1-D) and predictions are sequential (3-D),
+        # extract predictions at target position because Keras expects
+        # predictions and targets to have the same shape.
+        if getattr(prediction, "_keras_mask", None) is not None:
+            rank_check = tf.logical_and(
+                tf.equal(tf.rank(target), 1),
+                tf.equal(tf.rank(prediction), 3),
+            )
+            prediction = tf.cond(
+                rank_check, lambda: self._extract_masked_predictions(prediction), lambda: prediction
+            )
+
+        # Ensuring targets are one-hot encoded if they are not
+        target = tf.cond(
+            tf.rank(target) == tf.rank(prediction) - 1,
+            lambda: tf.one_hot(
+                tf.cast(target, tf.int32),
+                tf.shape(prediction)[-1],
+                dtype=prediction.dtype,
+            ),
+            lambda: tf.reshape(target, tf.shape(prediction)),
+        )
+
+        return prediction, target
+
+    def _adjust_ragged_predictions_and_targets(
+        self,
+        prediction: TensorLike,
+        target: tf.Tensor,
+    ):
+        """Adjusts the predictions and targets to ensure compatibility with most Keras losses and metrics.
+        This methods applies the following transformations to the target and prediction tensors:
+        - Select ragged targets based on the mask information, if defined.
+        - Remove mask information from the ragged targets and predictions.
+        - One-hot encode targets if their rank is one less than the rank of predictions.
+        - Ensure that targets have the same shape and dtype as predictions.
+        """
+        if isinstance(target, tf.RaggedTensor) and getattr(target, "_keras_mask", None) is not None:
+            # Select targets at masked positions and return
+            # a ragged tensor.
+            target = tf.ragged.boolean_mask(
+                target.with_row_splits_dtype(target.dtype), target._keras_mask
+            )
+
+        # Ensuring targets and preds have the same dtype
+        target = tf.cast(target, prediction.dtype)
+
+        # If targets are scalars (1-D) and predictions are sequential (3-D),
+        # extract predictions at target position because Keras expects
+        # predictions and targets to have the same shape.
+        if getattr(prediction, "_keras_mask", None) is not None:
+            rank_check = tf.logical_and(
+                tf.equal(tf.rank(target), 1),
+                tf.equal(tf.rank(prediction), 3),
+            )
+            prediction = tf.cond(
+                rank_check, lambda: self._extract_masked_predictions(prediction), lambda: prediction
+            )
+
+        # Take the flat values of predictions and targets as Keras
+        # losses does not support RaggedVariantTensor on GPU:
+        prediction = prediction.flat_values
```

Comment:
I think that by using `flat_values` for both `predictions` and `target` we misalign them with the shape of `sample_weight`. In that case, I think we should also flatten `sample_weight` if it has the same rank ?

Reply:
I had a similar [question]( https://github.com/NVIDIA-Merlin/models/pull/981#discussion_r1114558548 ) but couldn’t really figure out where to transform the `sample_weights`. One possible solution could be to add sample_weight as an additional argument to the `adjust_predictions_and_targets` function and apply the same transformations to align it with predictions and targets. What do you think?